# 1- Import Libraries

In [3]:
import pandas as pd
import numpy as np
import os

In [4]:
orders = pd.read_csv('olist_orders_dataset.csv')
customers = pd.read_csv('olist_customers_dataset.csv')
order_items = pd.read_csv('olist_order_items_dataset.csv')
products = pd.read_csv('olist_products_dataset.csv')
translation = pd.read_csv('product_category_name_translation.csv')
order_payments = pd.read_csv('olist_order_payments_dataset.csv')
sellers = pd.read_csv('olist_sellers_dataset.csv')
order_reviews = pd.read_csv('olist_order_reviews_dataset.csv')
geolocation = pd .read_csv('olist_geolocation_dataset.csv')

# 3- Data Exploration

# (1) Orders Table

In [76]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [77]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [78]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [79]:
orders.duplicated().sum()

np.int64(0)

In [80]:
orders.nunique()

order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

Orders Table - Observations
---------------------------
- The dataset contains 99,441 orders.
- All columns are stored as object data type, including date columns.
  These columns need to be converted to datetime for analysis.
- There are missing values in:
     order_approved_at (160)
     order_delivered_carrier_date (1783)
     order_delivered_customer_date (2965)
- Key Distinction:
     order_id: A unique identifier for each order (Primary Key of this table).
     customer_id: A foreign key that links each order to a specific customer in the Customers table. One customer can have multiple orders (one-to-      many relationship).  
- Missing values likely indicate: Orders that are not delivered yet or canceled orders
- No duplicate rows found in the dataset.
- The table contains multiple timestamps representing different stages of the order lifecycle.

# (2) Customers Table

In [81]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [82]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [83]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [84]:
customers.duplicated().sum()

np.int64(0)

In [85]:
customers.nunique()

customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

Customers Table - Observations
--------------------------------
- The dataset contains 99,441 customers.
- All columns are stored as object data types, except for customer_zip_code_prefix, which is an integer.
- There are NO missing values in any of the columns (100% complete data).
- Key Distinction: - customer_id: A unique key for each order (links to the Orders table).
- customer_unique_id: A unique identifier for the actual customer (useful for tracking repeat purchases).
- No duplicate rows found in the dataset.
- The table provides geographic distribution data (customer_city, customer_state) which is essential for regional sales analysis.

# (3) Order_Items Table

In [86]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [87]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [88]:
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [89]:
order_items.duplicated().sum()

np.int64(0)

In [90]:
order_items.nunique()

order_id               98666
order_item_id             21
product_id             32951
seller_id               3095
shipping_limit_date    93318
price                   5968
freight_value           6999
dtype: int64

In [91]:
order_items[['price', 'freight_value']].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


In [92]:
print("Zero prices:", (order_items['price'] == 0).sum())
print("Zero freight:", (order_items['freight_value'] == 0).sum())

Zero prices: 0
Zero freight: 383


In [93]:
print("Negative prices:", (order_items['price'] < 0).sum())

Negative prices: 0


Order Items Table - Observations
---------------------------------
- The dataset contains 112,650 entries, representing the individual products within each order.

- The dataset consists of 7 columns, including identifiers, pricing, and shipping information.

- Data Types: price and freight_value are stored as float64 (numeric).

   order_item_id is an int64.

- All other columns, including shipping_limit_date, are stored as object data types.

- There are NO missing values in the dataset (0 nulls across all columns).

- No duplicate rows found in the dataset.

- The shipping_limit_date column needs to be converted to a datetime format for any time-based shipping analysis.

- No zero or negative prices were found, indicating clean pricing data.

- However, prices range from 0.85 to 6,735 with a std higher than the mean, 
  suggesting a right-skewed distribution with high-value outliers that may 
  affect averages in future analysis.

- 383 orders have freight_value = 0, possibly indicating free shipping 
  promotions or data entry errors — worth investigating during analysis.

# (4) Products Table

In [94]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [95]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [96]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [97]:
products.duplicated().sum()

np.int64(0)

In [98]:
products.nunique()

product_id                    32951
product_category_name            73
product_name_lenght              66
product_description_lenght     2960
product_photos_qty               19
product_weight_g               2204
product_length_cm                99
product_height_cm               102
product_width_cm                 95
dtype: int64

Products Table - Observations
-----------------------------------
- The dataset contains 32,951 unique products.

- The table consists of 9 columns, covering product categories, descriptions, and physical dimensions.

- Data Types: 7 columns are stored as float64 (including dimensions and quantities)
2 columns (product_id, product_category_name) are stored as object data types.

- Missing Values Detected: 610 missing values in product_category_name, product_name_lenght, product_description_lenght, and product_photos_qty. This  suggests these products were listed without complete descriptive metadata.
- 2 missing values in physical dimensions (weight, length, height, width).
- Key Distinction: product_id is the Primary Key of this table. Each product has exactly one unique product_id.
- No duplicate rows found in the dataset.
- The product_category_name is in Portuguese, which may require mapping to English using the product_category_name_translation table for better reporting later.
- Physical Dimensions: All dimensions (weight, length, height, width) are available for almost all products, which is vital for logistics and freight cost analysis.

# (5) Translation Table

In [99]:
translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [100]:
translation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [101]:
translation.isnull().sum()



product_category_name            0
product_category_name_english    0
dtype: int64

In [102]:
translation.duplicated().sum()

np.int64(0)

Product Category Name Translation Table - Observations
-------------------------------------------------------
- The dataset contains 71 unique category mappings.
- The table consists of 2 columns: product_category_name (in Portuguese) and product_category_name_english (the corresponding English translation).
- Data Quality: - NO missing values (0 nulls). Both Portuguese and English names are fully populated for all 71 entries.
- No duplicate rows found in the dataset.
- Data Types: Both columns are stored as object (string) data types.
- Strategic Utility: This is a lookup table. It will be essential for merging with the products table to convert Portuguese category names into English, making the final analysis and visualizations more accessible to a global audience.
- Consistency Check: While this table has 71 categories, the products table might contain categories not listed here. This should be verified during the merging phase.

# (6) Order_Payments Table

In [5]:
order_payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [104]:
order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [105]:
order_payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [106]:
order_payments.duplicated().sum()

np.int64(0)

In [107]:
order_payments.nunique()

order_id                99440
payment_sequential         29
payment_type                5
payment_installments       24
payment_value           29077
dtype: int64

Order Payments Table - Observations
-------------------------------------
- The dataset contains 103,886 entries, detailing how customers paid for their orders.
- The table consists of 5 columns, providing information on payment methods, installments, and transaction values.
- Data Types: payment_value is stored as float64 (decimal numbers).
payment_sequential and payment_installments are stored as int64 (whole numbers).
order_id and payment_type are stored as object data types.
- Data Quality: - NO missing values (0 nulls). All payment records are fully populated.
- No duplicate rows found in the dataset.
- Payment Flexibility: The payment_installments column suggests that Olist offers a "Buy Now, Pay Later" option, which is a key insight for analyzing consumer behavior.
- Sequential Payments: The payment_sequential column indicates that a single order can be paid using multiple payment methods (e.g., part credit card, part voucher).
- Financial Integrity: This table is the primary source for calculating Total Revenue and understanding the distribution of payment methods (Credit Card, Boleto, Voucher, Debit Card).

# (7) Sellers Table

In [108]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [109]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [110]:
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [111]:
sellers.duplicated().sum()

np.int64(0)

In [112]:
sellers.nunique()

seller_id                 3095
seller_zip_code_prefix    2246
seller_city                611
seller_state                23
dtype: int64

Sellers Table - Observations
------------------------------
- The dataset contains 3,095 unique sellers registered on the Olist platform.
- The table consists of 4 columns, primarily focusing on the identification and geographic location of the sellers.
- Data Types: seller_zip_code_prefix is stored as an int64.
All other columns (seller_id, seller_city, seller_state) are stored as object data types.
- Data Quality: - NO missing values (0 nulls). The dataset is 100% complete.
- Key Distinction: seller_id is the Primary Key of this table. Each seller has exactly one unique seller_id.
- No duplicate rows found in the dataset.
- Geographic Distribution: Similar to the customers table, this dataset provides the city and state for each seller, which is essential for analyzing the Logistics Flow (calculating the distance between sellers and customers).
- Seller Concentration: This table allows for identifying which regions (states/cities) act as the primary hubs for sellers in the Brazilian market.

# (8) Order_Reviews Table

In [113]:
order_reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [114]:
order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [115]:
order_reviews.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [116]:
order_reviews.duplicated().sum()

np.int64(0)

In [117]:
order_reviews.nunique()

review_id                  98410
order_id                   98673
review_score                   5
review_comment_title        4527
review_comment_message     36159
review_creation_date         636
review_answer_timestamp    98248
dtype: int64

Order Reviews Table - Observations
-------------------------------------
- The dataset contains 99,224 entries, representing customer feedback and ratings for the orders.
- The table consists of 7 columns, capturing both numerical scores and text-based comments.
- Data Types: review_score is stored as an int64 (ranging from 1 to 5).
All other columns, including the dates, are stored as object data types.
- Significant Missing Values: - 87,656 missing values in review_comment_title.
58,247 missing values in review_comment_message.
- This indicates that the majority of customers provide a numerical rating without leaving a written title or comment.
- No duplicate rows found in the dataset.
- Temporal Data: The columns review_creation_date and review_answer_timestamp need to be converted to datetime format to analyze response times and trends in satisfaction over time.
- Sentiment Analysis Potential: Despite the missing values, there are over 40,000 comments available for text mining and sentiment analysis to understand the reasons behind low or high ratings.

# (9) Geolocation Table

In [118]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [119]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [120]:
geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [121]:
geolocation.duplicated().sum()

np.int64(261831)

In [122]:
geolocation.nunique()

geolocation_zip_code_prefix     19015
geolocation_lat                717360
geolocation_lng                717613
geolocation_city                 8011
geolocation_state                  27
dtype: int64

Geolocation Table - Observations
---------------------------------
- The dataset is the largest in the project, containing 1,000,163 entries.
- The table consists of 5 columns, providing the geographic coordinates (latitude and longitude) for Brazilian zip codes.
- Data Types: geolocation_lat and geolocation_lng are stored as float64.
geolocation_zip_code_prefix is an int64.
geolocation_city and geolocation_state are object data types.
- Missing Values: NO missing values (0 nulls). The dataset is geographically complete.
- High Redundancy (Critical): The dataset contains 261,831 duplicate rows. This is a very high number of duplicates that must be handled (removed) to avoid unnecessary processing load and to ensure accuracy in spatial calculations.
- Purpose: This table acts as a Geospatial Reference. It allows for mapping the locations of customers and sellers, which is vital for calculating shipping distances and analyzing regional market penetration.
- Cleaning Recommendation: Given the high number of duplicates and the fact that one zip code might have multiple coordinate points, it is recommended to "de-duplicate" or group the data by zip code before merging it with other tables.

# 4- Data Cleaning

# (1) Date Conversion

In [123]:
# date conversion for orders table
date_columns = ['order_purchase_timestamp', 'order_approved_at', 
                'order_delivered_carrier_date', 'order_delivered_customer_date', 
                'order_estimated_delivery_date']
for col in date_columns:
    orders[col] = pd.to_datetime(orders[col])

In [124]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


In [125]:
# date conversion for order_reviews table
order_reviews['review_creation_date'] = pd.to_datetime(order_reviews['review_creation_date'])
order_reviews['review_answer_timestamp'] = pd.to_datetime(order_reviews['review_answer_timestamp'])

In [126]:
order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  object        
 1   order_id                 99224 non-null  object        
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  object        
 4   review_comment_message   40977 non-null  object        
 5   review_creation_date     99224 non-null  datetime64[ns]
 6   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB


In [127]:
# date conversion for order_items table
order_items['shipping_limit_date'] = pd.to_datetime(order_items['shipping_limit_date'])

In [128]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  object        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  object        
 3   seller_id            112650 non-null  object        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 6.0+ MB


# (2) Handling Nulls

In [129]:
# In the products table, we found 610 missing values in the names. Instead of deleting them, 
# we will replace them with the word 'Unknown' to ensure that the associated sales data is not lost.

products['product_category_name'] = products['product_category_name'].fillna('Unknown')

In [130]:
# Replace missing numerical values (such as weight and dimensions) with median.

cols_to_fix = ['product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']
for col in cols_to_fix:
    products[col] = products[col].fillna(products[col].median())

In [131]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32951 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32951 non-null  float64
 6   product_length_cm           32951 non-null  float64
 7   product_height_cm           32951 non-null  float64
 8   product_width_cm            32951 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


Key Takeaways for Cleaning Nulls Phase:
--------------------------------------

- The "Big Clean": The Geolocation table is the only one with a significant duplicate issue that needs to be addressed programmatically to reduce the data size.

- Missing Metadata: We will need to replace the 610 missing values ​​in the Products table with the keyword 'Unknown' before merging and compiling.

- Logical Nulls: The missing values ​​in the Orders table are not data "errors," but rather represent the order journey (orders that have not yet been delivered). Therefore, we will handle them carefully when calculating delivery times.

- Textual Nulls: In the Reviews table, we will focus on the review_score as quantitative data and treat the missing comments as empty messages that do not affect the accuracy of the numerical rating.


# (3) Removing Duplicates

In [132]:
# The geolocation table contains over 260,000 duplicate rows. We will only keep the unique version for each postal code.
# Delete all duplicate rows.

geolocation = geolocation.drop_duplicates()

In [133]:
# To ensure one copy for each postal code.

geolocation = geolocation.drop_duplicates(subset=['geolocation_zip_code_prefix'], keep='first')

In [134]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19015 entries, 0 to 999846
Data columns (total 5 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   geolocation_zip_code_prefix  19015 non-null  int64  
 1   geolocation_lat              19015 non-null  float64
 2   geolocation_lng              19015 non-null  float64
 3   geolocation_city             19015 non-null  object 
 4   geolocation_state            19015 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 891.3+ KB


# (4) Unifying and correcting textual data (Merging Translation)

In [135]:
# We will merge the translation table with the products table to get the names in English directly.

products = pd.merge(products, translation, on='product_category_name', how='left')

In [136]:
# If we find products that were not in the translation table,
# we put their original name or Unknown.

products['product_category_name_english'] = products['product_category_name_english'].fillna(products['product_category_name'])


In [137]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_category_name_english
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0,perfumery
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0,art
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0,sports_leisure
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0,baby
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0,housewares


# (5) Calculating "Actual Delivery Time"

In [138]:
# To make sure that the columns are in datetime format

orders[['order_purchase_timestamp', 'order_approved_at', 
                'order_delivered_carrier_date', 'order_delivered_customer_date', 
                'order_estimated_delivery_date']].dtypes



order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object

In [139]:
# Calculate the actual delivery time in days

orders['actual_delivery_days'] = (orders['order_delivered_customer_date'] - orders['order_purchase_timestamp']).dt.days

In [140]:
# Calculating the difference between expected and actual delivery
# Estimated vs Actual (If positive then the delivery was ahead of schedule. if negativethen there was a delay)

orders['delivery_diff_days'] = (orders['order_estimated_delivery_date'] 
                                - orders['order_delivered_customer_date']).dt.days

In [141]:
# Displaying the new columns

orders[['actual_delivery_days','delivery_diff_days']].head()

,actual_delivery_days,delivery_diff_days
0,8.0,7.0
1,13.0,5.0
2,9.0,17.0
3,13.0,12.0
4,2.0,9.0


In [142]:
# Displaying descriptive statistics for new periods

orders[['actual_delivery_days', 'delivery_diff_days']].describe()

,actual_delivery_days,delivery_diff_days
count,96476.000000,96476.000000
mean,12.094086,10.876881
std,9.551746,10.183854
min,0.000000,-189.000000
25%,6.000000,6.000000
50%,10.000000,11.000000
75%,15.000000,16.000000
max,209.000000,146.000000


# 5- Export Clean Data to CSV

In [143]:
os.makedirs('olist_clean', exist_ok=True)

orders.to_csv('olist_clean/orders_clean.csv', index=False)
customers.to_csv('olist_clean/customers_clean.csv', index=False)
order_items.to_csv('olist_clean/order_items_clean.csv', index=False)
products.to_csv('olist_clean/products_clean.csv', index=False)
order_payments.to_csv('olist_clean/order_payments_clean.csv', index=False)
sellers.to_csv('olist_clean/sellers_clean.csv', index=False)
order_reviews.to_csv('olist_clean/order_reviews_clean.csv', index=False)
geolocation.to_csv('olist_clean/geolocation_clean.csv', index=False)

print("Done! All files saved in 'olist_clean' folder")


Done! All files saved in 'olist_clean' folder
